cdc/entry/schema/snapshot.go

```diff
@@ -0,0 +1,1151 @@
+	tag := negative(s.currentTs)
+	start := newVersionedEntityName(-1, schema, tag)
+	end := newVersionedEntityName(-1, schema, negative(uint64(0)))
+	s.schemaNameToID.AscendRange(start, end, func(i btree.Item) bool {
+		id = i.(versionedEntityName).target
+		ok = id >= 0 // negative values are treated as invalid.
+		return false
+	})
+	return
+}
+
+// TableIDByName returns the tableID by table schemaName and tableName.
+// The second returned value is false if no table with the specified name is found.
+func (s *Snapshot) TableIDByName(schema string, table string) (int64, bool) {
+	s.rwlock.RLock()
+	defer s.rwlock.RUnlock()
+	return s.tableIDByNameLocked(schema, table)
+}
+
+func (s *Snapshot) tableIDByNameLocked(schema string, table string) (id int64, ok bool) {
+	var prefix int64
+	prefix, ok = s.schemaIDByNameLocked(schema)
+	if ok {
+		tag := negative(s.currentTs)
+		start := newVersionedEntityName(prefix, table, tag)
+		end := newVersionedEntityName(prefix, table, negative(uint64(0)))
+		s.tableNameToID.AscendRange(start, end, func(i btree.Item) bool {
+			id = i.(versionedEntityName).target
+			ok = id >= 0 // negative values are treated as invalid.
+			return false
+		})
+	}
+	return
+}
+
+// TableByName queries a table by name,
+// The second returned value is false if no table with the specified name is found.
+// NOTE: The returned table info should always be READ-ONLY!
+func (s *Snapshot) TableByName(schema, table string) (*model.TableInfo, bool) {
+	s.rwlock.RLock()
+	defer s.rwlock.RUnlock()
+	return s.tableByNameLocked(schema, table)
+}
+
+func (s *Snapshot) tableByNameLocked(schema, table string) (info *model.TableInfo, ok bool) {
+	id, ok := s.tableIDByNameLocked(schema, table)
+	if !ok {
+		return nil, ok
+	}
+	return s.physicalTableByIDLocked(id)
+}
+
+// SchemaByTableID returns the schema ID by table ID.
+func (s *Snapshot) SchemaByTableID(tableID int64) (*timodel.DBInfo, bool) {
+	s.rwlock.RLock()
+	defer s.rwlock.RUnlock()
+	tableInfo, ok := s.physicalTableByIDLocked(tableID)
+	if !ok {
+		return nil, false
+	}
+	return s.schemaByIDLocked(tableInfo.SchemaID)
+}
+
+// IsTruncateTableID returns true if the table id have been truncated by truncate table DDL.
+func (s *Snapshot) IsTruncateTableID(id int64) bool {
+	s.rwlock.RLock()
+	defer s.rwlock.RUnlock()
+	tag, ok := s.tableTagByID(id, true)
+	return ok && s.truncatedTables.Get(newVersionedID(id, tag)) != nil
+}
+
+// IsIneligibleTableID returns true if the table is ineligible.
+func (s *Snapshot) IsIneligibleTableID(id int64) bool {
+	s.rwlock.RLock()
+	defer s.rwlock.RUnlock()
+	return s.isIneligibleTableIDLocked(id)
+}
+
+func (s *Snapshot) isIneligibleTableIDLocked(id int64) (ok bool) {
+	tag, ok := s.tableTagByID(id, false)
+	return ok && s.ineligibleTables.Get(newVersionedID(id, tag)) != nil
+}
+
+func (s *Snapshot) tableTagByID(id int64, nilAcceptable bool) (foundTag uint64, ok bool) {
+	s.rwlock.RLock()
+	defer s.rwlock.RUnlock()
```

Comment:
I'm afraid mixing locked API and unlocked API can easily causing dead lock.

```go
func (s *Snapshot) IsTruncateTableID(id int64) bool {
	s.rwlock.RLock()
	defer s.rwlock.RUnlock()
	tag, ok := s.tableTagByID(id, true)
}
...

func (s *Snapshot) tableTagByID(id int64, nilAcceptable bool) (foundTag uint64, ok bool) {
	s.rwlock.RLock()
	defer s.rwlock.RUnlock()
```

----

 Maybe we can separate thread-safe API and thread-unsafe API into two structs.

```go
// All thread-safe APIs go here
type Snapshot struct {
  mu sync.Mutex
  s snapshot
}

// -----

// All thread-unsafe APIs go here.
type snapshot struct {
}
```